lib/eventsource/index.js

```diff
@@ -0,0 +1,448 @@
+        // value other than CLOSED, sets the readyState attribute to CLOSED
+        // and fires an event named error at the EventSource object. Once the
+        // user agent has failed the connection, it does not attempt to
+        // reconnect.
+        if (response.aborted) {
+          this.close()
+          this.dispatchEvent(new ErrorEvent('error', { error: response.error }))
+          return
+          // 2. Otherwise, if res is a network error, then reestablish the
+          // connection, unless the user agent knows that to be futile, in
+          // which case the user agent may fail the connection.
+        } else {
+          this.#reconnect()
+          return
+        }
+        // 3. Otherwise, if res's status is not 200, [...], then fail the
+        // connection.
+      } else if (response.status !== 200) {
+        let message = `Unexpected status code: ${response.status}`
+
+        if (response.status === 204) {
+          message = 'No content'
+        }
+        this.close()
+        this.dispatchEvent(new ErrorEvent('error', { message }))
+        return
+      }
+
+      // 3. Otherwise, [...] if res's
+      // `Content-Type` is not `text/event-stream`, then fail the
+      // connection.
+
+      const contentType = response.headersList.get('content-type', true)
+      const mimeType = contentType !== null ? parseMIMEType(contentType) : 'failure'
+      if (mimeType === 'failure' || mimeType.essence !== 'text/event-stream') {
+        this.close()
+        this.dispatchEvent(new ErrorEvent('error', { message: 'Invalid content-type' }))
+        return
+      }
+      // 4. Otherwise, announce the connection and interpret res's body
+      // line by line.
+
+      // When a user agent is to announce the connection, the user agent
+      // must queue a task which, if the readyState attribute is set to a
+      // value other than CLOSED, sets the readyState attribute to OPEN
+      // and fires an event named open at the EventSource object.
+      // @see https://html.spec.whatwg.org/multipage/server-sent-events.html#sse-processing-model
+      this.#readyState = OPEN
+      this.dispatchEvent(new Event('open', {}))
+
+      const eventSourceStream = new EventSourceStream({
+        eventSourceState: this.#settings,
+        push: (eventPayload) => {
+          this.dispatchEvent(new MessageEvent(
+            eventPayload.type,
+            eventPayload.payload
+          ))
+        }
+      })
+
+      pipeline(response.body.stream,
+        eventSourceStream,
+        (error) => {
+          if (
+            error &&
+            error.aborted === false
+          ) {
+            this.close()
+            this.dispatchEvent(new ErrorEvent('error', { error }))
+          }
+        })
+    }
+
+    fetchParam.timingInfo = createOpaqueTimingInfo({})
+
+    fetchParam.controller = new Fetch(getGlobalDispatcher())
+
+    this.#controller = fetchParam.controller
+
+    await mainFetch(fetchParam)
```

Comment:
Is there a reason that you using mainFetch over `fetching`? This doesn't seem correct to me.

Reply:
fetching will create the fetchParams for you and return the controller. No need to do it yourself!